### PR TITLE
Rey

### PR DESCRIPTION
**Pull request by Roxxistic**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,13 +92,14 @@ Search is performed by Alpha-Beta-Pruning at a deepness of 4.
...
 - Position of one pawn relative to the game board:
-  - The further advanced a pawn, the better.
+  - The further advanced a pawn, the better. _(implemented)_
   - A pawn that has an unblocked path to the finish line
-    - is great.
-    - is less great, if there is an opponent's pawn in a neighbouring line up ahead and thus could kill our pawn.
+    - is great. _(implemented)_
+    - is less great, if there is an opponent's pawn in a neighbouring line up ahead and thus could kill our pawn. _(not implemented)_
 - Position of one pawn relative to our and opponent's other pawns:
+  _(not implemented since this is covered by the possible moves in the next level in the game tree and the count of pawns)_
```
Comment (Tartori) [inline README.md]: remove not implemented stuff

File: README.md [hunk 2]
```diff
@@ -92,13 +92,14 @@ Search is performed by Alpha-Beta-Pruning at a deepness of 4.
...
-  - The more pawns one has compared to the number of pawns the opponent has, the better.
+  - The more pawns one has compared to the number of pawns the opponent has, the better. _(implemented)_
 - Position of one pawn relative to the game board:
-  - The further advanced a pawn, the better.
+  - The further advanced a pawn, the better. _(implemented)_
   - A pawn that has an unblocked path to the finish line
-    - is great.
-    - is less great, if there is an opponent's pawn in a neighbouring line up ahead and thus could kill our pawn.
+    - is great. _(implemented)_
+    - is less great, if there is an opponent's pawn in a neighbouring line up ahead and thus could kill our pawn. _(not implemented)_
```
Comment (Tartori) [inline README.md]: remove not implemented information

File: README.md [hunk 3]
```diff
@@ -92,13 +92,14 @@ Search is performed by Alpha-Beta-Pruning at a deepness of 4.
...
 - Total pawns:
-  - The more pawns one has compared to the number of pawns the opponent has, the better.
+  - The more pawns one has compared to the number of pawns the opponent has, the better. _(implemented)_
 - Position of one pawn relative to the game board:
-  - The further advanced a pawn, the better.
+  - The further advanced a pawn, the better. _(implemented)_
   - A pawn that has an unblocked path to the finish line
-    - is great.
-    - is less great, if there is an opponent's pawn in a neighbouring line up ahead and thus could kill our pawn.
+    - is great. _(implemented)_
```
Comment (Tartori) [inline README.md]: remove not implemented information

File: README.md [hunk 4]
```diff
@@ -92,13 +92,14 @@ Search is performed by Alpha-Beta-Pruning at a deepness of 4.
 The heuristic evaluation function is based on various considerations:
 
 - Total pawns:
-  - The more pawns one has compared to the number of pawns the opponent has, the better.
+  - The more pawns one has compared to the number of pawns the opponent has, the better. _(implemented)_
 - Position of one pawn relative to the game board:
-  - The further advanced a pawn, the better.
+  - The further advanced a pawn, the better. _(implemented)_
```
Comment (Tartori) [inline README.md]: remove not implemented information

File: README.md [hunk 5]
```diff
@@ -92,13 +92,14 @@ Search is performed by Alpha-Beta-Pruning at a deepness of 4.
 The heuristic evaluation function is based on various considerations:
 
 - Total pawns:
-  - The more pawns one has compared to the number of pawns the opponent has, the better.
+  - The more pawns one has compared to the number of pawns the opponent has, the better. _(implemented)_
```
Comment (Tartori) [inline README.md]: remove not implemented information

File: controllers/machine_ab_pruning_actor.py [hunk 6]
```diff
@@ -47,5 +47,12 @@ def __find_best_solution_rec(self, depth, game, alpha=-100):
                 alpha = sol
         return alpha, alphapawn, alphamove
 
-    def __get_heuristic(self, game):
+    def _get_heuristic(self, game):
         raise NotImplementedError("Abstract method.")
+
+    def _get_fields_delta(self, board, current_player):
```
Comment (Tartori) [inline controllers/machine_ab_pruning_actor.py:53]: rename to _get_pawn_delta

File: controllers/machine_complex_heuristic_actor.py [hunk 7]
```diff
@@ -11,6 +13,57 @@ class MachineComplexHeuristicActor(MachineABPruningActor):
     def __init__(self, name: str, state, playerId):
         super().__init__(name, state, playerId)
 
-    def __get_heuristic(self, game):
+    def _get_heuristic(self, game):
         # TODO: implement heuristic!!!
```
Comment (Tartori) [inline controllers/machine_complex_heuristic_actor.py]: remove todo

**Review by Tartori — CHANGES_REQUESTED** [review]

File: controllers/machine_complex_heuristic_actor.py [hunk 8]
```diff
@@ -11,6 +13,57 @@ class MachineComplexHeuristicActor(MachineABPruningActor):
...
+                    adv_board[r][c] *= (size - 1 - r - 1)**2
+
+        adv_delta = self._get_fields_delta(adv_board, game._current_player)
+
+        # A pawn that has an unblocked path to the finish line is great
+        board = game._get_bord()
+        top_field_value = [0] * size
+        btm_field_value = [0] * size
+
+        for r in range(size):
```
Comment (Tartori) [inline controllers/machine_complex_heuristic_actor.py]: only iterate once

File: controllers/machine_complex_heuristic_actor.py [hunk 9]
```diff
@@ -11,6 +13,57 @@ class MachineComplexHeuristicActor(MachineABPruningActor):
...
+                    # squared weight
+                    adv_board[r][c] *= (r - 1)**2
+                elif adv_board[r][c] == Player.USER:
+                    # best row is at the "top" of the board
+                    adv_board[r][c] *= (size - 1 - r - 1)**2
+
+        adv_delta = self._get_fields_delta(adv_board, game._current_player)
+
+        # A pawn that has an unblocked path to the finish line is great
+        board = game._get_bord()
```
Comment (Tartori) [inline controllers/machine_complex_heuristic_actor.py:34]: use adv_board for this to get increased score the further a pawn has progressed especially if his sight is free.

File: controllers/machine_complex_heuristic_actor.py [hunk 10]
```diff
@@ -11,6 +13,57 @@ class MachineComplexHeuristicActor(MachineABPruningActor):
...
+        for r in range(size):
+            for c in range(size):
+                if adv_board[r][c] == Player.OPP:
+                    # best row is at the "bottom" of the board
+                    # -1 to prevent the first move to always be two steps.
+                    # squared weight
+                    adv_board[r][c] *= (r - 1)**2
+                elif adv_board[r][c] == Player.USER:
+                    # best row is at the "top" of the board
+                    adv_board[r][c] *= (size - 1 - r - 1)**2
```
Comment (Tartori) [inline controllers/machine_complex_heuristic_actor.py]: use *= (size - 1 - r - 1) * (size - 1 - r - 1) instead, more performant

File: controllers/machine_complex_heuristic_actor.py [hunk 11]
```diff
@@ -11,6 +13,57 @@ class MachineComplexHeuristicActor(MachineABPruningActor):
...
+        adv_board = deepcopy(game._get_bord())
+        size = game._size
+
+        for r in range(size):
+            for c in range(size):
+                if adv_board[r][c] == Player.OPP:
+                    # best row is at the "bottom" of the board
+                    # -1 to prevent the first move to always be two steps.
+                    # squared weight
+                    adv_board[r][c] *= (r - 1)**2
```
Comment (Tartori) [inline controllers/machine_complex_heuristic_actor.py]: use *= (r-1)*(r-1) instead, more performant

**Review by Tartori — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (Roxxistic) [reply README.md]: included but separated and explained why not implemented.

File: README.md (hunk 2 above)
Comment (Roxxistic) [reply README.md]: included but separated and explained why not implemented.

File: README.md (hunk 3 above)
Comment (Roxxistic) [reply README.md]: included but separated and explained why not implemented.

File: README.md (hunk 4 above)
Comment (Roxxistic) [reply README.md]: included but separated and explained why not implemented.

File: README.md (hunk 5 above)
Comment (Roxxistic) [reply README.md]: included but separated and explained why not implemented.

File: controllers/machine_ab_pruning_actor.py (hunk 6 above)
Comment (Roxxistic) [reply controllers/machine_ab_pruning_actor.py:53]: correct naming but hat to adjust comment. its not only the count of pawns.

File: controllers/machine_complex_heuristic_actor.py (hunk 7 above)
Comment (Roxxistic) [reply controllers/machine_complex_heuristic_actor.py]: done

File: controllers/machine_complex_heuristic_actor.py (hunk 8 above)
Comment (Roxxistic) [reply controllers/machine_complex_heuristic_actor.py]: somewhat done. has to have iteration from both sides.

File: controllers/machine_complex_heuristic_actor.py (hunk 9 above)
Comment (Roxxistic) [reply controllers/machine_complex_heuristic_actor.py:34]: done

File: controllers/machine_complex_heuristic_actor.py (hunk 10 above)
Comment (Roxxistic) [reply controllers/machine_complex_heuristic_actor.py]: done

File: controllers/machine_complex_heuristic_actor.py (hunk 11 above)
Comment (Roxxistic) [reply controllers/machine_complex_heuristic_actor.py]: done

**Comment by Roxxistic** [comment]
I improved the complex heuristic and completed the documentation.
However gameplay currently neither works in master nor in this branch anymore.
Might you take a look? @Tartori

File: controllers/machine_complex_heuristic_actor.py [hunk 12]
```diff
@@ -1,16 +1,108 @@
...
+        usr_now_blocked = [False] * size
+        opp_now_blocked = [False] * size
+
+        # Traversal of board backwards for performance reasons.
+        # (free paths flags)
+        # Of course this could also be done by flipping calculation of
+        # the row indices. But that seems counterintuitive.
+        for r in range(size - 1, -1, -1):
+            for c in range(size):
+                # Row indices for both perspectives.
```
Comment (Tartori) [inline controllers/machine_complex_heuristic_actor.py:56]: nice

File: controllers/machine_complex_heuristic_actor.py [hunk 13]
```diff
@@ -1,16 +1,108 @@
...
+                    # - the further advanced (given as value in r_usr),
+                    #   the better.
+                    # - if the column ahead is free from Player.OPP's pawns,
+                    #   gets a bonus.
+                    # To prevent each pawn from taking 2 fields as a first
+                    # step, subtracted 1 from value.
+                    adv_board[r_usr][c] *= (r_usr - 1) * (r_usr - 1)
+
+                    if not usr_now_blocked[c]:
+                        adv_board[r_usr][c] *= 10  # TODO: choose best weight
```
Comment (Tartori) [inline controllers/machine_complex_heuristic_actor.py:81]: maybe put in different weight? but probably fine

**Review by Tartori — APPROVED** [review]

**Pull request merged**